src/back/hlsl/writer.rs

```diff
@@ -1856,14 +1856,54 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                     }
                 };
 
-                let var_handle = self.fill_access_chain(module, pointer, func_ctx)?;
-                // working around the borrow checker in `self.write_expr`
-                let chain = mem::take(&mut self.temp_access_chain);
-                let var_name = &self.names[&NameKey::GlobalVariable(var_handle)];
+                let accessing_byte_addr_buf = {
+                    fn is_atomic_scalar_ty(module: &Module, ty: Handle<crate::Type>) -> bool {
+                        match &module.types[ty].inner {
+                            &TypeInner::Atomic {
+                                kind: crate::ScalarKind::Sint | crate::ScalarKind::Uint,
+                                ..
+                            } => true,
+                            &TypeInner::Array { base, .. } | &TypeInner::Pointer { base, .. } => {
+                                is_atomic_scalar_ty(module, base)
+                            }
+                            _ => false,
+                        }
+                    }
+
+                    fn is_atomic_scalar_expr(
+                        module: &Module,
+                        func_ctx: &back::FunctionCtx,
+                        expr: Handle<crate::Expression>,
+                    ) -> bool {
+                        match &func_ctx.expressions[expr] {
+                            &crate::Expression::GlobalVariable(handle) => {
+                                is_atomic_scalar_ty(module, module.global_variables[handle].ty)
+                            }
+                            &crate::Expression::Access { base, .. }
+                            | &crate::Expression::AccessIndex { base, .. } => {
+                                is_atomic_scalar_expr(module, func_ctx, base)
+                            }
+                            _ => false,
```

Comment:
question: RE: the wildcard pattern lines a la `_ => false`: Are there any expression or type constructs that we need in this PR before merging?